src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/DSASignatureDeformatter.cs

```diff
@@ -0,0 +1,67 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Internal.Cryptography;
+
+namespace System.Security.Cryptography
+{
+    public class DSASignatureDeformatter : AsymmetricSignatureDeformatter
+    {
+        DSA _dsaKey;
+        string _algName;
+
+        public DSASignatureDeformatter()
+        {
+            // The hash algorithm default is SHA1
+            _algName = HashAlgorithmNames.SHA1;
+        }
+
+        public DSASignatureDeformatter(AsymmetricAlgorithm key) : this()
+        {
+            if (key == null)
+                throw new ArgumentNullException(nameof(key));
+
+            _dsaKey = (DSA)key;
+        }
+
+        public override void SetKey(AsymmetricAlgorithm key)
+        {
+            if (key == null)
+                throw new ArgumentNullException(nameof(key));
+
+            _dsaKey = (DSA)key;
+        }
+
+        public override void SetHashAlgorithm(string strName)
+        {
+            // The implemenation is symmetric with RSAPKCS1SignatureDeformatter even though _algName
+            // is not used in VerifySignature.
+            try
+            {
+                // Verify the name
+                Oid.FromFriendlyName(strName, OidGroup.HashAlgorithm);
+
+                // Keep the raw strName as Oid may change the case ("SHA1" to "sha1") making it incompatible.
+                _algName = strName;
```

Comment:
I didn't see anything in FIPS-186-3 that we need to do here, other than don't throw if "SHA256" for example is used, just for the sake of having coherent API semantics. Today in desktop for example you are forced to pass in "SHA1" but since it's not actually used later, you can still pass in a hash from SHA256 successfully. So what I did in latest code is just throw CryptographicUnexpectedException if it doesn't match any valid hash algorithm.